### PR TITLE
`<fstream>`: `fstream::close` shouldn't corrupt heap (after `putback`)

### LINKED ISSUE #342: <fstream>: Closing a filebuf after putback falls back to pbackfail leads to heap corruption
The code at the bottom triggers the behavior.

The issue is caused by putback running over the start of the buffer (enforced by `seekg`) and hence calling `pbackfail` which puts the char into an internal buffer and sets that as the new buffer. The pointer to that buffer is stored in the backing structure of a `FILE*` where normally a heap-allocated buffer is stored. On calling `fclose` (through the streams close function) the implementation tries to deallocate the buffer without realizing that it is not a heap allocated buffer.

In Debug builds this triggers an assertion but on release builds it silently frees a stack pointer possibly leading to a potential vulnerability.

Depending on the memory manager used this might be used to exploit code that uses or can be forced to use that putback sequence.

```
#include <fstream>

int main()
{
    {
        std::ofstream f("test.txt");
        f << "ab";
    }
    std::fstream f("test.txt");
    f.seekg(1);
    f.get(); // b
    f.putback('b');
    f.putback('a');
    f.close();
    return 0;
}
```

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #342

### CODE REVIEW

File: tests/std/tests/GH_000342_filebuf_close/test.cpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#include <fstream>
+
+using namespace std;
+
+int main() {
+    {
+        ofstream f("test.txt");
+        f << "ab";
+    }
+    fstream f("test.txt");
```
Comment (AlexGuteniev) [inline tests/std/tests/GH_000342_filebuf_close/test.cpp:13]: Almost-hiding. Maybe another variable name?

File: tests/std/tests/GH_000342_filebuf_close/test.cpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    f.seekg(1);
+    f.get(); // b
```
Comment (AlexGuteniev) [inline tests/std/tests/GH_000342_filebuf_close/test.cpp:15]: care to `assert` that?

**Review by AlexGuteniev — APPROVED** [review]

File: stl/inc/fstream [hunk 3]
```diff
@@ -176,10 +176,6 @@ public:
     }
 
     __CLR_OR_THIS_CALL ~basic_filebuf() noexcept override {
-        if (_Myfile) {
-            _Reset_back(); // revert from _Mychar buffer
-        }
```
Comment (CaseyCarter) [inline stl/inc/fstream:181]: I see how the addition on 381 fixes the bug, but I'm concerned that removing `_Reset_back()` here is creating a different bug for users of the `basic_filebuf(FILE*)` extension constructor. That constructor sets `_Myfile` but not `_Closef`, so if `_Set_back()` was ever evaluated I think we'll leave the `FILE` in a bad state after destroying `this` without evaluating `_Reset_back()`.

Am I missing something?

Comment (fsb4000) [reply stl/inc/fstream:181]: I missed that there is a constructor that does not close the file :(

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this long-standing bug! :file_cabinet: :lady_beetle: :smile_cat:

**Pull request merged**